src/cargo/core/compiler/build_runner/compilation_files.rs

```diff
@@ -677,12 +676,45 @@ fn compute_metadata(
                 .target_config(CompileKind::Host)
                 .links_overrides
                 != unit.links_overrides;
-        target_configs_are_different.hash(&mut hasher);
+        target_configs_are_different.hash(&mut shared_hasher);
     }
 
+    let mut c_metadata_hasher = shared_hasher.clone();
+    // Mix in the target-metadata of all the dependencies of this target.
+    let mut dep_c_metadata_hashes = deps_metadata
+        .iter()
+        .map(|m| m.c_metadata)
+        .collect::<Vec<_>>();
+    dep_c_metadata_hashes.sort();
+    dep_c_metadata_hashes.hash(&mut c_metadata_hasher);
+
+    let mut c_extra_filename_hasher = shared_hasher.clone();
+    // Mix in the target-metadata of all the dependencies of this target.
+    let mut dep_c_extra_filename_hashes = deps_metadata
+        .iter()
+        .map(|m| m.c_extra_filename)
+        .collect::<Vec<_>>();
+    dep_c_extra_filename_hashes.sort();
+    dep_c_extra_filename_hashes.hash(&mut c_extra_filename_hasher);
+    // Avoid trashing the caches on RUSTFLAGS changing via `c_extra_filename`
+    //
+    // Limited to `c_extra_filename` to help with reproducible build / PGO issues.
```

Comment:
> This should (probably) avoid the PGO problems, since the symbols won't change. However, this could still cause problems with reproducible builds if the rlib filename somehow leaks into the resulting binary (I don't have a reason to think it does, but it seems risky).

> I believe this does because the filenames of object files show up as the codegen unit names in dwarf debug info (and probably PDB?). I suspect that everything else in the binary will be the same, but builds differing only in extra-filename are likely to still produce different binaries.

Have we checked if it doesn't affect PGO?
We would like to check if anything is leaking rlib filename in debug info files.

Reply:
Any idea how to go about that?  I have no experience with PGO and we have no tests focusing on it.

Reply:
I'll do some experiments this week and see if we can have tests written down!

Reply:
Posted https://github.com/rust-lang/cargo/pull/14859

Reply:
The test passes with this PR for me

Reply:
For most rustflags, changing them changes the binary anyway. Remap is special in that it takes a build-specific value and removes it, making builds more reproducible.

Reply:
Do you suggest that we consider `--remap-path-refix` case as unsupported?

Reply:
... we generate debug files per code-gen unit?  Yikes

This isn't present for `cargo b --config 'profile.dev.strip="debuginfo"'` but I assume that is not a relevant case for build reproducibility folks.  Are `dwp` files considered important for those cases?

Anyways, some options for moving this forward
- Move forward as-is, requiring build-reproducibility people to use the unstable feature
- Block until the unstable feature is stable
- Parse rustc flags enough to ignore `--remap-path-prefix`, like #6966
- Assume even less knowledge of rustc flags and don't include RUSTFLAGS in `-Cextra-filename` if `--remap-path-prefix` is present, meaning people using it don't get this benefit but most people will get the benefit

Thoughts?

Reply:
@epage the conclusion from Monday's meeting is that we're going to reintroduce #6966 along with this PR, right?

Reply:
In the meeting notes, it says we were going to move forward with the hack where we conditionally add RUSTFLAGS to `-Cextra-filename`

Reply:
Thanks. I am forgetting things more often recently…
Given that, this PR is ready to go!

Reply:
I removed this from the queue as I'd like to add that hack before this gets merged so we don't accidentally break people.

Reply:
> Will move forward with ed's hack.


I am confused by the meeting note. I assumed Ed's hack is this PR. The “hack” you are going to add is #6966, no?

Reply:
#6966 filters out `--remap-path-prefix` from being hashed.

The hack I proposed is that we skip hashing *all* RUSTFLAGS if something that looks like `--remap-path-prefix` is present.